---
title: Consolidating DB analyses for CHART-seq
author: Aaron Lun
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    fig_caption: false
    toc_float: yes
---

```{r, echo=FALSE, results="hide"}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
```

# Reading in the results

Here we consolidate the various window-based analyses into a single set of results.
First, we define the files to load in:

```{r}
library(csaw)
all.files <- list.files(full=TRUE, pattern="^out_.*rds$")
all.files
```

And we load in the results:

```{r}
all.results <- lapply(all.files, readRDS)
all.regions <- lapply(all.results, "[[", i="windows")
all.dres <- lapply(all.results, "[[", i="results")
```

# Reporting DB regions

For interpretation, regions are more relevant than windows.
To get this, we cluster windows into regions using a simple single-linkage algorithm.
Windows less than 200 bp apart are merged together, with regions allowed to be 5 kbp in size.

```{r}
merged <- consolidateWindows(all.regions, 
    merge.args=list(tol=200, max.width=5e3))
summary(as.integer(table(unlist(merged$id))))
merged$region
```

We get the differential statistics for each merged region.
The _p_-value for each region is calculated from the _p_-values of the constituent windows.
We also compute an "empirical FDR", based on the number of regions with best log-fold changes going in the wrong direction (i.e., enriched in sense).
This provides more control of biological false positives due to non-specific enrichment.

```{r}
tabemp <- consolidateTests(merged$id, all.dres, 
    merged$weight, FUN=empiricalFDR)
head(tabemp)
summary(tabemp$FDR <= 0.05)
```

We can then compute the number and direction of the significant regions.

```{r}
table(tabemp$direction[tabemp$FDR <= 0.05]) 
```

Within each region, we also get the coordinates of the window with the lowest _p_-value.
This is useful for pinpointing where the strongest (differential) binding is occuring within large regions.

```{r}
tabbest <- consolidateTests(merged$id, all.dres,
    merged$weight, FUN=getBestTest)
head(tabbest)
```

Adding annotation.

```{r}
library(org.Hs.eg.db)
library(EnsDb.Hsapiens.v86)
fiddling <- merged$region
seqlevels(fiddling) <- paste0("tmp_", seqlevels(fiddling)) # avoid clash with yeast X.
seqlevels(fiddling) <- sub("^tmp_chr", "", seqlevels(fiddling))
suppressWarnings({
    surrounds <- detailRanges(orgdb=org.Hs.eg.db, txdb=EnsDb.Hsapiens.v86, 
        fiddling, key.field="ENSEMBL")
})
```

We assemble all of these statistics into a table for convenient examination.

```{r}
output <- data.frame(as.data.frame(merged$region)[,1:3], tabemp, 
    best.logFC=tabbest$logFC, surrounds)
output <- output[order(output$PValue),]    
write.table(file="db_regions.tsv", output,
    row.names=FALSE, sep="\t", quote=FALSE)        
head(output)
```

```{r, echo=FALSE, results="hide"}
gc()
```

# Characterizing the binding profile

We explore whether the binding sites are likely to occur in promoters, exons, introns or intergenic regions.
We take all binding sites at a FDR of 30% (which is large, but the data are simply that noisy).

```{r}
sig.sites <- merged$region[tabemp$FDR <= 0.3]
seqlevels(sig.sites) <- sub("^chr", "", seqlevels(sig.sites))

prom.reg <- promoters(EnsDb.Hsapiens.v86, upstream=3000, downstream=500)
suppressWarnings(olap.prom <- overlapsAny(sig.sites, prom.reg))
sum(olap.prom)

exon.reg <- exons(EnsDb.Hsapiens.v86)
suppressWarnings(olap.exon <- overlapsAny(sig.sites, exon.reg))
sum(olap.exon)

# Loading human enhancers
library(rtracklayer)
library(BiocFileCache)
bfc <- BiocFileCache("../../data_store", ask = FALSE)
enh.loc <- bfcrpath(bfc, "http://enhanceratlas.org/data/enhseq/human/Hela.fasta")
enh.reg <- GRanges(sub("_.*", "", names(import(enh.loc))))
seqlevels(enh.reg) <- sub("^chr", "", seqlevels(enh.reg))    
suppressWarnings(olap.enh <- overlapsAny(sig.sites, enh.reg))
sum(olap.enh)

intron.reg <- intronicParts(EnsDb.Hsapiens.v86)
suppressWarnings(olap.intron <- overlapsAny(sig.sites, intron.reg))

remaining <- !logical(length(sig.sites))
collected <- list()
collected$Promoter <- sum(remaining & olap.prom)
remaining <- remaining & !olap.prom
collected$Exon <- sum(remaining & olap.exon)
remaining <- remaining & !olap.exon
collected$Enhancer <- sum(remaining & olap.enh)
remaining <- remaining & !olap.enh
collected$Intron <- sum(remaining & olap.intron)
remaining <- remaining & !olap.intron
collected$Intergenic <- sum(remaining)
DataFrame(collected)
```

# Reporting DB genes

## Defining the regions

We obtain genic regions from the Ensembl annotation:

```{r}
hs.genes <- genes(EnsDb.Hsapiens.v86)
hs.genes[,2]
```

We also kick out each gene by 3 kbp at the 5' end, to capture binding at the putative promoeter region.

```{r}
suppressWarnings({
    hs.genes <- resize(hs.genes, width(hs.genes)+3e3, fix="end")
})
hs.genes[,2]
```

Fixing the sequence levels:

```{r}
seqlevels(hs.genes) <- paste0("chr", seqlevels(hs.genes))
head(seqlevels(hs.genes))
```

Getting the gene symbols.

```{r}
anno <- data.frame(Ensembl=hs.genes$gene_id, Symbol=hs.genes$gene_name)
head(anno)   
```

## Computing gene-level statistics

Consolidating results across genes:

```{r}
suppressWarnings({
    molaps <- consolidateWindows(all.regions, region=hs.genes)
})
```

We get the empirical FDR for each gene.

```{r}
tabemp <- consolidateOverlaps(molaps$olap, all.dres, 
    molaps$weight, FUN=empiricalOverlaps)
head(tabemp)
summary(tabemp$FDR <= 0.05)
```

We can then compute the number and direction of the significant genes.

```{r}
table(tabemp$direction[tabemp$FDR <= 0.05]) 
```

We also compute the best statistics for each gene.

```{r}
tabbest <- consolidateOverlaps(molaps$olap, all.dres,
	molaps$weight, FUN=getBestOverlaps)
head(tabbest)
```

We save this result to file.

```{r}
output <- data.frame(anno, tabemp, best.logFC=tabbest$logFC) 
output <- output[order(output$PValue),]    
output <- output[!is.na(output$PValue),]
write.table(file="db_genes.tsv", output,
    row.names=FALSE, sep="\t", quote=FALSE)
head(output)
```

# Session information

We finish up by reporting the session information.

```{r}
sessionInfo()
```
