---
title: Semi-quantitative proteomics analysis upon lncRNA pulldown 
author: Aaron Lun
date: "`r Sys.Date()`"
output:
  html_document:
    fig_caption: false
    toc: true
    toc_float: true
    depth: 3
    number_sections: true
    theme: united 
    highlight: tango 
---

```{r, echo=FALSE, results="hide"}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
dir.create("results-poisson", showWarning=FALSE)
dir.create("results-logit", showWarning=FALSE)
options(width=100)
```

# Introduction

Proteomics data generated by Lovorka Stojic, after pulling down HeLa lysate with lncRNA, scrambled control or nothing.
We use the number of unique peptides detected as a proxy for the abundance of the protein.
First we load in all the counts, which requires some finesse to handle Excel files.

```{r}
library(readxl)
all.files <- list.files("data", recursive=TRUE, pattern="xlsx", full=TRUE)
collected <- accessions <- vector("list", length(all.files))
for (f in seq_along(all.files)) {
    incoming <- read_excel(all.files[f])
    collected[[f]] <- incoming[,c("Accession", "# Unique Peptides")]
    accessions[[f]] <- incoming[,c("Accession", "Description")]
}
```

We identify the unique accessions across the entire data set.

```{r}
accessions <- do.call(rbind, accessions)
accessions <- accessions[!duplicated(accessions$Accession),]
nrow(accessions)
```

We assemble a count matrix using these accessions.

```{r}
mat <- matrix(0L, nrow(accessions), length(collected))
for (f in seq_along(collected)) {
    current <- collected[[f]]
    m <- match(current$Accession, accessions$Accession)
    mat[m,f] <- current[,"# Unique Peptides",drop=TRUE]
}
rownames(mat) <- accessions$Accession
```

We also designate metadata for each sample.

```{r}
groups <- c("Beads", "C1_full", "C1_scr", "C1_ex2", "C1_ex13", 
                "271_full", "271_scr", "271_ex1", "271_ex24")
g <- factor(groups[as.integer(sub(".*_([0-9]+)[A-Z]\\.xlsx", "\\1", all.files))])
batch <- factor(sub(".*_[0-9]+([A-Z])\\.xlsx", "\\1", all.files))
colnames(mat) <- paste0(g, batch)
head(data.frame(all.files, g, batch))
```

# Testing for differences in unique proteins with _edgeR_

## Setting up the analysis

Constructing a `DGEList` object and a design matrix.

```{r}
library(edgeR)
y <- DGEList(mat, genes=accessions)
design <- model.matrix(~0 + g + batch)
```

Setting the library sizes to be constant because it doesn't make sense to normalize by the total number of unique peptides in a qualitative setting.
For example, beads have a lot lower number of peptides detected, which is consistent with reduced pulldown.

```{r}
y$samples$lib.size <- 1e6
barplot(colSums(mat), las=2, ylab="Total number of peptides")
```

We filter out proteins that are detected with fewer than 5 peptides across all libraries.

```{r}
y <- y[rowSums(y$counts) >= 4,]
dim(y)
```

We estimate the NB dispersions.

```{r}
y <- estimateDisp(y, design)
plotBCV(y)
```

This motivates a Poisson model for the unique peptide counts.

```{r}
fit <- glmFit(y, design, dispersion=0)
```

## Testing for proteins enriched in 271 pulldown

### Full pulldown vs scrambled

Setting up the contrast.

```{r}
con <- makeContrasts(g271_full - g271_scr, levels=design)
show.libs <- grep("271_(full|scr)", colnames(mat))
out.file <- file.path("results-poisson", "271-full_vs_scr.tsv")
```

Running the contrast and saving the results.
This involves some finesse to only look for genes with positive log-fold changes (enriched _against_ control).

```{r runcon}
res <- glmLRT(fit, contrast=con)
z <- sqrt(pmax(0, res$table$LR)) * sign(res$table$logFC)
res$table$PValue <- pnorm(z, lower.tail=FALSE)
summary(decideTestsDGE(res))
best <- topTags(res, n=Inf)$table
best <- cbind(best, mat[best$Accession, show.libs])
head(best)
write.table(best, file=out.file, sep="\t", quote=FALSE, row.names=FALSE)
```

### Full pulldown vs beads 

Setting up the contrast.

```{r}
con <- makeContrasts(g271_full - gBeads, levels=design)
show.libs <- grep("(271_full|Beads)", colnames(mat))
out.file <- file.path("results-poisson", "271-full_vs_beads.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon"}
```

### Exon 1 pulldown vs scrambled

Setting up the contrast.

```{r}
con <- makeContrasts(g271_ex1 - g271_scr, levels=design)
show.libs <- grep("271_(ex1|scr)", colnames(mat))
out.file <- file.path("results-poisson", "271-ex1_vs_scr.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon"}
```

### Exon 1 pulldown vs beads 

Setting up the contrast.

```{r}
con <- makeContrasts(g271_ex1 - gBeads, levels=design)
show.libs <- grep("(271_ex1|Beads)", colnames(mat))
out.file <- file.path("results-poisson", "271-ex1_vs_beads.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon"}
```

### Exon 2-4 pulldown vs scrambled

Setting up the contrast.

```{r}
con <- makeContrasts(g271_ex24 - g271_scr, levels=design)
show.libs <- grep("271_(ex24|scr)", colnames(mat))
out.file <- file.path("results-poisson", "271-ex24_vs_scr.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon"}
```

### Exon 2-4 pulldown vs beads 

Setting up the contrast.

```{r}
con <- makeContrasts(g271_ex24 - gBeads, levels=design)
show.libs <- grep("(271_ex24|Beads)", colnames(mat))
out.file <- file.path("results-poisson", "271-ex24_vs_beads.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon"}
```

## Testing for proteins enriched in C1 pulldown

### Full pulldown vs scrambled

Setting up the contrast.

```{r}
con <- makeContrasts(gC1_full - gC1_scr, levels=design)
show.libs <- grep("C1_(full|scr)", colnames(mat))
out.file <- file.path("results-poisson", "C1-full_vs_scr.tsv")
```

Running the contrast and saving the results.

```{r runcon}
```

### Full pulldown vs beads 

Setting up the contrast.

```{r}
con <- makeContrasts(gC1_full - gBeads, levels=design)
show.libs <- grep("(C1_full|Beads)", colnames(mat))
out.file <- file.path("results-poisson", "C1-full_vs_beads.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon"}
```

### Exon 2 pulldown vs scrambled

Setting up the contrast.

```{r}
con <- makeContrasts(gC1_ex2 - gC1_scr, levels=design)
show.libs <- grep("C1_(ex2|scr)", colnames(mat))
out.file <- file.path("results-poisson", "C1-ex2_vs_scr.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon"}
```

### Exon 2 pulldown vs beads 

Setting up the contrast.

```{r}
con <- makeContrasts(gC1_ex2 - gBeads, levels=design)
show.libs <- grep("(C1_ex2|Beads)", colnames(mat))
out.file <- file.path("results-poisson", "C1-ex2_vs_beads.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon"}
```

### Exon 1-3 pulldown vs scrambled

Setting up the contrast.

```{r}
con <- makeContrasts(gC1_ex13 - gC1_scr, levels=design)
show.libs <- grep("C1_(ex13|scr)", colnames(mat))
out.file <- file.path("results-poisson", "C1-ex13_vs_scr.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon"}
```

### Exon 1-3 pulldown vs beads 

Setting up the contrast.

```{r}
con <- makeContrasts(gC1_ex13 - gBeads, levels=design)
show.libs <- grep("(C1_ex13|Beads)", colnames(mat))
out.file <- file.path("results-poisson", "C1-ex13_vs_beads.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon"}
```

# Testing for presence or absence with logistic regression

## Setting up 

We consider 1 unique peptide as a threshold for detection.

```{r}
detected <- mat >= 1
barplot(colSums(detected), las=2, ylab="Number of detected proteins")
```

We set up a function to do the logistic regression.

```{r}
logitTest <- function(detected, design, contrast) { 
    contrast <- as.matrix(contrast)
    qrc <- qr(contrast)
    ncontrasts <- qrc$rank
    coef <- 1:ncontrasts
    contrast <- drop(contrast)
    Dvec <- rep.int(1, nrow(design))
    Dvec[coef] <- diag(qrc$qr)[coef]
    Q <- qr.Q(qrc, complete = TRUE, Dvec = Dvec)
    design <- design %*% Q
    design0 <- design[,-coef]

    ngenes <- nrow(mat)
    all.p <- all.coef <- numeric(ngenes)
    for (g in seq_len(ngenes)) { 
        curd <- detected[g,]
        full <- glm.fit(x=design, curd, family=binomial())
        null <- glm.fit(x=design0, curd, family=binomial()) 
        all.p[g] <- null$deviance - full$deviance
        all.coef[g] <- full$coefficients[coef]        
    }
    all.z <- sqrt(pmax(0, all.p)) * sign(all.coef)
    all.p <- pnorm(all.z, lower=FALSE)
    return(data.frame(DeltaLOD=all.coef, PValue=all.p, FDR=p.adjust(all.p, method="BH")))
}
```

## Testing for proteins enriched in 271 pulldown

### Full pulldown vs scrambled

Setting up the contrast.

```{r}
con <- makeContrasts(g271_full - g271_scr, levels=design)
show.libs <- grep("271_(full|scr)", colnames(mat))
out.file <- file.path("results-logit", "271-full_vs_scr.tsv")
```

Running the contrast and saving the results.

```{r runcon2}
res <- logitTest(detected, design, con)
sum(res$FDR <= 0.05)
res <- data.frame(accessions, res, mat[,show.libs], check.names=FALSE)
res <- res[order(res$PValue),]
head(res)
write.table(res, file=out.file, sep="\t", quote=FALSE, row.names=FALSE)
```

### Full pulldown vs beads 

Setting up the contrast.

```{r}
con <- makeContrasts(g271_full - gBeads, levels=design)
show.libs <- grep("(271_full|Beads)", colnames(mat))
out.file <- file.path("results-logit", "271-full_vs_beads.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon2"}
```

### Exon 1 pulldown vs scrambled

Setting up the contrast.

```{r}
con <- makeContrasts(g271_ex1 - g271_scr, levels=design)
show.libs <- grep("271_(ex1|scr)", colnames(mat))
out.file <- file.path("results-logit", "271-ex1_vs_scr.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon2"}
```

### Exon 1 pulldown vs beads 

Setting up the contrast.

```{r}
con <- makeContrasts(g271_ex1 - gBeads, levels=design)
show.libs <- grep("(271_ex1|Beads)", colnames(mat))
out.file <- file.path("results-logit", "271-ex1_vs_beads.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon2"}
```

### Exon 2-4 pulldown vs scrambled

Setting up the contrast.

```{r}
con <- makeContrasts(g271_ex24 - g271_scr, levels=design)
show.libs <- grep("271_(ex24|scr)", colnames(mat))
out.file <- file.path("results-logit", "271-ex24_vs_scr.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon2"}
```

### Exon 2-4 pulldown vs beads 

Setting up the contrast.

```{r}
con <- makeContrasts(g271_ex24 - gBeads, levels=design)
show.libs <- grep("(271_ex24|Beads)", colnames(mat))
out.file <- file.path("results-logit", "271-ex24_vs_beads.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon2"}
```

## Testing for proteins enriched in C1 pulldown

### Full pulldown vs scrambled

Setting up the contrast.

```{r}
con <- makeContrasts(gC1_full - gC1_scr, levels=design)
show.libs <- grep("C1_(full|scr)", colnames(mat))
out.file <- file.path("results-logit", "C1-full_vs_scr.tsv")
```

Running the contrast and saving the results.

```{r runcon2}
```

### Full pulldown vs beads 

Setting up the contrast.

```{r}
con <- makeContrasts(gC1_full - gBeads, levels=design)
show.libs <- grep("(C1_full|Beads)", colnames(mat))
out.file <- file.path("results-logit", "C1-full_vs_beads.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon2"}
```

### Exon 2 pulldown vs scrambled

Setting up the contrast.

```{r}
con <- makeContrasts(gC1_ex2 - gC1_scr, levels=design)
show.libs <- grep("C1_(ex2|scr)", colnames(mat))
out.file <- file.path("results-logit", "C1-ex2_vs_scr.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon2"}
```

### Exon 2 pulldown vs beads 

Setting up the contrast.

```{r}
con <- makeContrasts(gC1_ex2 - gBeads, levels=design)
show.libs <- grep("(C1_ex2|Beads)", colnames(mat))
out.file <- file.path("results-logit", "C1-ex2_vs_beads.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon2"}
```

### Exon 1-3 pulldown vs scrambled

Setting up the contrast.

```{r}
con <- makeContrasts(gC1_ex13 - gC1_scr, levels=design)
show.libs <- grep("C1_(ex13|scr)", colnames(mat))
out.file <- file.path("results-logit", "C1-ex13_vs_scr.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon2"}
```

### Exon 1-3 pulldown vs beads 

Setting up the contrast.

```{r}
con <- makeContrasts(gC1_ex13 - gBeads, levels=design)
show.libs <- grep("(C1_ex13|Beads)", colnames(mat))
out.file <- file.path("results-logit", "C1-ex13_vs_beads.tsv")
```

Running the contrast and saving the results.

```{r ref.label="runcon2"}
```


# Wrapping up

```{r}
saveRDS(file="objects.rds", y)
sessionInfo()
```
