---
title: Comparing RNA-seq and CHART-seq results
author: Aaron Lun
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc_float: yes
    fig_caption: false
---

```{r, echo=FALSE}
knitr::opts_chunk$set(error=FALSE, warning=FALSE, message=FALSE)
```

# Reading in the data

This intersects the differential expression results from RNA-seq with the #_MARK_# differential binding results.
First we read in the DE result tables.

```{r}
rna.path <- "../rnaseq/analysis/results_lfc"
rna.res <- read.delim(file.path(rna.path, "cons_271_all.txt"), 
    row.names=1, stringsAsFactors=FALSE)
```

... and to the cut-and-run differential binding tables for each gene:

```{r}
chart.path <- "../chartseq/analysis/db_genes.tsv"
chart.res <- read.delim(chart.path, row.names=1, stringsAsFactors=FALSE)
```

Also obtaining the `DGEList` object for more detailed testing.

```{r}
library(edgeR)
y <- readRDS("../rnaseq/analysis/object.rds")
design <- y$design
```

# Intersecting all results

Figuring out the intersection.
We don't restrict the universe here, as genes that don't exhibit a different in the CHART-seq (against control) are clearly not bound at all.
Note that we can't use intersection-union tests due to the use of the empirical FDR, so we'll just do it in a more _ad hoc_ manner.

```{r}
sig.chart <- rownames(chart.res)[chart.res$FDR <= 0.3]
sig.rna <- rownames(rna.res)[rna.res$adj.P.Val <= 0.05] 
intersect(sig.rna, sig.chart) # In RNAi+LNA and CHART 
length(setdiff(sig.rna, sig.chart)) # In RNAi+LNA but not CHART
length(setdiff(sig.chart, sig.rna)) # In CHART but not RNAi+LNA
```

Unfortunately, it is not generally possible to split the Venn into LNA and RNAi-specific sections.
This is because the Venn has no ability to handle genes that are significant but in different directions between LNA and RNAi.
As a result, there would always be a minor discrepancy in the sum of the numbers in the intersections compared to the actual total number of DE genes for each technology.

# Gene set testing on CHART-bound sites

Throwing in a few ROAST tests based on the direction of the change.

```{r}
indices <- list(Bound=sig.chart)

# Versus Ambion 
con <- makeContrasts(
    gRNA_interference.wild_type_genotype.271_siRNA.batch_1
        - gRNA_interference.wild_type_genotype.Ambion_control.batch_1,
    levels=design)
roast(y, indices, design=design, contrast=con)

# Versus Dharmacon
con <- makeContrasts(
    gRNA_interference.wild_type_genotype.271_siRNA.batch_1
        - gRNA_interference.wild_type_genotype.Dharmacon_control.batch_1,
    levels=design)
roast(y, indices, design=design, contrast=con)

# Versus control A
con <- makeContrasts(
    gLNA.wild_type_genotype.271_LNA.batch_3
        - gLNA.wild_type_genotype.Negative_control_A.batch_3,
    levels=design)
roast(y, indices, design=design, contrast=con)

# Versus control B
con <- makeContrasts(
    gLNA.wild_type_genotype.271_LNA.batch_3
        - gLNA.wild_type_genotype.Negative_control_B.batch_3,
    levels=design)
roast(y, indices, design=design, contrast=con)
```

# Wrapping up

```{r}
sessionInfo()
```    
